main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -116,4 +116,96 @@ object SafetyError {
+        |
+        |${code(loc, "the illegal derivation occurs here.")}
+        |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = None
+  }
+
+  /**
+    * An error raised to indicate an invalid `this` parameter for a method.
+    *
+    * @param clazz The Java interface being implemented.
+    * @param name  The name of the method with the invalid `this` parameter.
+    * @param loc   The source location of the method.
+    */
+  case class InvalidThis(clazz: java.lang.Class[_], name: String, loc: SourceLocation) extends SafetyError {
+    def summary: String = s"Invalid `this` parameter for method '${name}'."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+        |>> Invalid `this` parameter for method '${red(name)}''.
+        |
+        |${code(loc, "the method occurs here.")}
+        |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = Some({
+      s"""
+        | The first argument to any method must be `this`, and must have the same type as the interface
+        | being implemented.
+        |""".stripMargin
+    })
+  }
+
+  /**
+    * An error raised to indicate an unimplemented Java interface method
+    *
+    * @param clazz The Java interface being implemented.
+    * @param name  The name of the unimplemented method.
+    * @param loc   The source location of the object derivation.
+    */
+  case class UnimplementedMethod(clazz: java.lang.Class[_], name: String, loc: SourceLocation) extends SafetyError {
+    def summary: String = s"No implementation found for method '${name}' of interface '${clazz.getName}'."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+        |>> No implementation found for method '${red(name)}' of interface '${red(clazz.getName)}'.
+        |
+        |${code(loc, "the object occurs here.")}
+        |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = None
+  }
+
+  /**
+    * An error raised to indicate that an object derivation includes a method that doesn't exist
+    * in the Java interface being implemented.
+    *
+    * @param clazz The Java interface being implemented.
+    * @param name  The name of the extra method.
+    * @param loc   The source location of the method.
+    */
+  case class ExtraMethod(clazz: java.lang.Class[_], name: String, loc: SourceLocation) extends SafetyError {
+    def summary: String = s"Method '${name}' not found in interface '${clazz.getName}'"
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+        |>> Method '${red(name)}' not found in interface '${red(clazz.getName)}'
+        |
+        |${code(loc, "the method occurs here.")}
+        |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = None
+  }
```

Comment:
Suggestions for additional errors:

- MissingDefaultConstructor -- the class lacks a zero arg constructor.
- IllegalFinalClass -- the class is final.
- IllegalFinalMethod - the method exists and is final.
- IllegalNonPublicClassOrInterface -- the class or interface is not public.
- IllegalNonPublicConstructor -- the constructor is not public.